src/main/java/org/dsa/iot/alarm/AlarmStreamer.java

```diff
@@ -0,0 +1,171 @@
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Constants
+    ///////////////////////////////////////////////////////////////////////////
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Fields
+    ///////////////////////////////////////////////////////////////////////////
+
+    private boolean closed = false;
+    private Collection container;
+    private AlarmCursor initialSet;
+    private ActionResult request;
+    private Table table;
+    private LinkedList<AlarmRecord> updates = new LinkedList<>();
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Constructors
+    ///////////////////////////////////////////////////////////////////////////
+
+    /**
+     *  Will set this as the close handler on the given request and will add and
+     *  remove itself from the given container.
+     *
+     *  @param container Where to add and remove this instnace.
+     *  @param initialSet Optional.
+     */
+    public AlarmStreamer(
+            Collection container,
+            ActionResult request,
+            AlarmCursor initialSet) {
+        request.setCloseHandler(this);
+        this.container = container;
+        synchronized (container) {
+            container.add(this);
+        }
+        this.request = request;
+        this.initialSet = initialSet;
+        request.setStreamState(StreamState.INITIALIZED);
+        this.table = request.getTable();
+        table.setMode(Table.Mode.STREAM);
+        table.sendReady();
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Methods
+    ///////////////////////////////////////////////////////////////////////////
+
+    /**
+     * Force close the stream.
+     */
+    public void close() {
+        closed = true;
+    }
+
+    /**
+     * Does not return until there is a record, or the stream is closed.
+     *
+     * @return Possibly null if the stream is closed.
+     */
+    public AlarmRecord getNextUpdate() {
+        synchronized (updates) {
+            while (isValid() && (updates.size() == 0)) {
+                try {
+                    updates.wait(5000);
+                } catch (Exception ignore) {}
+                if (updates.size() > 0) {
+                    return updates.removeFirst();
+                }
+            }
+        }
+        return null;
+    }
+
+    /**
+     * True if there are pending updates.
+     */
+    public synchronized boolean hasUpdates() {
+        return updates.size() > 0;
+    }
+
+    /**
+     * True if both sides of the connection are open.
+     */
+    public boolean isValid() {
+        return isOpen() && !closed;
+    }
+
+    /**
+     * Sends the initial set of alarms if not null, then sends updates until the
+     * stream is closed.
+     */
+    public void run() {
+        table.waitForStream(10000, true);
```

Comment:
I would extract methods within this.. Seems to have a lot of work done in here, and it's hard to figure out without reading carefully.